configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java

```diff
@@ -533,10 +532,17 @@ public boolean delete(ApplicationId applicationId) {
             } else {
                 transaction.commit();
             }
-            return true;
         } finally {
             applicationTransaction.ifPresent(ApplicationTransaction::close);
         }
+
+        // Get the lock to remove the application. Other config servers should have attempted to
+        // get the lock (and lock is fair), so this will essentially let us wait for the others to finish.
+        // If some of them are down they will not be taking the lock, if something else is wrong, getting the
+        // lock times out after 1 minute
+        try (var lock = tenantApplications.lock(applicationId)) {
```

Comment:
Should we consider a barrier here?

Reply:
We are deleting an application and all nodes related to it in ZooKeeper, so there is nowhere to store it. We could store it somewhere else or add application state as we do for sessions. Makes it even more complicated, though. Can't say this is an elegant solution, either, though...